api/src/main/java/io/druid/segment/loading/BaseDataSegmentFinder.java

```diff
@@ -0,0 +1,74 @@
+/*
+ * Licensed to Metamarkets Group Inc. (Metamarkets) under one
+ * or more contributor license agreements. See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership. Metamarkets licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package io.druid.segment.loading;
+
+import io.druid.java.util.common.Pair;
+import io.druid.java.util.common.logger.Logger;
+import io.druid.timeline.DataSegment;
+
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+public abstract class BaseDataSegmentFinder implements DataSegmentFinder
+{
+  private static final Logger log = new Logger(BaseDataSegmentFinder.class);
+
+  @Override
+  public Set<DataSegment> findSegments(String workingDirPath, boolean updateDescriptor) throws SegmentLoadingException
+  {
+    final Map<String, Pair<DataSegment, Long>> timestampedSegments = new HashMap<>();
+
+    findSegments(workingDirPath, updateDescriptor, timestampedSegments);
+
+    return timestampedSegments.values().stream().map(x -> x.lhs).collect(Collectors.toSet());
+  }
+
+  /**
+   * Adds dataSegment if it does not exist in timestampedSegments. If it exists, replaces entry if segmentModifiedAt is
+   * newer than stored timestamp.
+   *
+   * @param timestampedSegments map of <segmentID, Pair<segment, modifiedAt>> containing segments with modified time
+   * @param dataSegment         segment to add
+   * @param segmentModifiedAt   segment modified timestamp
+   */
+  protected void putRetainingNewest(
```

Comment:
IMO, the implementations would be less convoluted if the helpers in this file were static utility methods rather than inherited superclass methods. I'm in the school of thought where inheritance is best used when nothing else really makes sense (neither composition nor utility methods).

In particular, I feel that the pattern in play here -- an abstract class implements a method from an interface and then creates a new method for its subclasses to implement -- is usually more difficult to understand than utility methods that can be called directly by the potential subclasses. It breaks the connection between the subclasses and the interface they implement, and requires the reader to understand the link established in the abstract class too.

I won't hold up the patch over this though.